File: python/05_print_linklist_reverse.py
class LinkNode:
	def __init__(self, value):
		self.element = value
		self.next = None

def append( pHead, value ):
	node = LinkNode(value)
	if pHead is None:
		pHead = node
	else:
		cur_node = pHead
		while cur_node.next is not None:
			cur_node = cur_node.next
		cur_node.next = node
	return pHead

def print_reverse( pHead ):
	node = pHead
	nodes_list = []
	while node is not None:
		nodes_list.append(node.element)
		node = node.next
	while len(nodes_list) > 0:
		print nodes_list.pop(),
	print ''

if __name__ == '__main__':
	pHead = None
	print_reverse( pHead )

	pHead = append( pHead, 4 )
	print_reverse( pHead )

	pHead = append( pHead, 5 )
	print_reverse( pHead )

	pHead = append( pHead, 3 )
	print_reverse( pHead )

File: cpp/05_print_linklist_reverse.cpp
#include <iostream>
#include <stack>

using namespace std;

struct LinkNode
{
	int        m_nKey;
	LinkNode * m_pNext;
};

void AddToTail(LinkNode ** pHead, int value)
{
	LinkNode * pNew = new LinkNode();
	pNew->m_nKey = value;
	pNew->m_pNext = NULL;

	if (*pHead == NULL)
		*pHead = pNew;
	else
	{
		LinkNode * pNode = *pHead;
		while (pNode->m_pNext != NULL)
			pNode = pNode->m_pNext;
		pNode->m_pNext = pNew;
	}
}

void PrintReverse(LinkNode * pHead)
{
	LinkNode * pNode = pHead;
	stack<int> nodes_stack;
	while (pNode != NULL)
	{
		nodes_stack.push(pNode->m_nKey);
		pNode = pNode->m_pNext;
	}
	while(!nodes_stack.empty())
	{
		cout << nodes_stack.top() << " ";
		nodes_stack.pop();
	}
	cout << endl;
}

int main()
{
	LinkNode * pHead = NULL;
	AddToTail(&pHead, 4);
	AddToTail(&pHead, 5);
	AddToTail(&pHead, 3);
	PrintReverse(pHead);
	return 0;
}